integration.py

```diff
@@ -0,0 +1,897 @@
+SAFE_TRANSFORMS = [Additivity, PullConstant, Expand, PolynomialDivision]
+
+TRIGFUNCTION_INTEGRALS = {
+    "sin": lambda x: -Cos(x),
+    "cos": Sin,
+    "tan": lambda x: -Log(Cos(x)),
+    "csc": lambda x: -Log(Csc(x) - Cot(x)),
+    "sec": lambda x: Log(Sec(x) + Tan(x)),
+    "cot": lambda x: Log(Sin(x)),
+}
+
+
+def _check_if_solvable(node: Node):
+    expr = node.expr
+    var = node.var
+    answer = None
+    if isinstance(expr, Power):
+        if expr.base == var and isinstance(expr.exponent, Const):
+            n = expr.exponent
+            answer = (1 / (n + 1)) * Power(var, n + 1) if n != -1 else Log(expr.base)
+        elif isinstance(expr.base, Symbol) and expr.base != var:
+            answer = expr * var
+
+    elif isinstance(expr, Symbol):
+        answer = Fraction(1 / 2) * Power(var, 2) if expr == var else expr * var
+    elif isinstance(expr, Const):
+        answer = expr * var
+    elif isinstance(expr, TrigFunction) and not expr.is_inverse:
+        answer = TRIGFUNCTION_INTEGRALS[expr.function](expr.inner)
+
+    if answer is None:
+        return
+
+    node.type = "SOLUTION"
+    node.solution = answer.simplify()
+
+
+def _cycle(node: Node) -> Optional[Union[Node, Literal["SOLVED"]]]:
+    # 1. APPLY ALL SAFE TRANSFORMS
+    _integrate_safely(node)
+
+    # now we have a tree with all the safe transforms applied
+    # 2. LOOK IN TABLE
+    for leaf in node.unfinished_leaves:
+        _check_if_solvable(leaf)
+
+    if len(node.unfinished_leaves) == 0:
+        return "SOLVED"
+
+    # 3. APPLY HEURISTICS
+    next_node = node.unfinished_leaves[0]  # random lol
+    _integrate_heuristically(next_node)
+
+    next_next_node = _get_next_node_post_heuristic(next_node)
+    return next_next_node
+
+
+def _get_next_node_post_heuristic(node: Node) -> Node:
+
+    if len(node.unfinished_leaves) == 0:
+        if node.is_failed:
+            # if on the first cycle, there are zero safe transforms AND no heuristics then
+            # node here would be the root.
+            if node.parent is None:
+                return None
+
+            # we want to go back and solve the parent
+            parent = node.parent
+            while len(parent.children) == 1 or parent.type == "AND":
+                if parent.parent is None:
+                    # we've reached root.
+                    # this means... we can't solve this integral.
+                    return None
+                parent = parent.parent
+            # now parent is the lowest OR node with multiple children
+            return _get_next_node_post_heuristic(parent)
+        else:
+            raise NotImplementedError("TODO _get_next_node for success nodes")
+
+    if len(node.unfinished_leaves) == 1:
+        return node.unfinished_leaves[0]
+
+    if len(node.unfinished_leaves) > 1:
+        return _nesting_node(node)
+
+
+# a recursive function.
+# find the simplest problem to work on that makes progress.
+def _nesting_node(node: Node) -> Node:
```

Comment:
We talked about it before but this might be a little overcomplicated. I still don't understand the utility of minimaxing here you need to explain to me again I think 😔

Reply:
yeah it might be overcomplicated.